Feat(preferences): Consolidated Preferences System: User, Collection, and Global Preferences Updates

Gitesh307 opened the pull request:
This combined PR unifies the work from issues #7445, #6843, #7440, and #7442 to deliver a complete overhaul of the preferences system across Specify.

It removes a number of obsolete remote preferences, migrates remaining tree-viewer configuration to the modern User Preferences framework, introduces a dedicated Collection Preferences interface for collection-level configuration, and adds a new Global Preferences page for institution-wide settings managed by Institution Administrators.

### Summary of Changes

**1. Removal of legacy remotePrefs and migration to User Preferences (Fixes #7445)**  
A set of outdated configuration keys from remotePrefs.ts and CollectionDefinitions.tsx has been retired. These keys—such as tree-viewer sort settings, rank thresholds, and TaxonTreeEditor.DisplayAuthor—either already existed in User Preferences or were no longer relevant. All code paths depending on these legacy keys were removed, and Tree Viewer now relies completely on the User Preferences system.

**2. Migration of TreeEditor.Rank.Threshold to User Preferences (Fixes #6843)**  
TreeEditor.Rank.Threshold no longer uses the legacy remote preference. Rank visibility and object count display are now fully controlled through User Preferences. Existing threshold values are not auto-migrated and can be updated manually per collection as needed.

**3. Introduction of the Collection Preferences UI (Fixes #7440)**  
A new, structured Collection Preferences page consolidates settings such as picklist scope, attachment defaults, tree management rules, statistics, Specify Network configuration, and catalog number inheritance. This page is role-gated, saved per collection, and uses updated localization handling to address scanner issues. No database schema changes were required.

Screenshots:
<img width="350" height="845" alt="image" src="https://github.com/user-attachments/assets/751fc33f-5924-4664-a730-8e8e6f559586" />
<img width="700" height="1521" alt="image" src="https://github.com/user-attachments/assets/194bc3c9-dd90-425c-a40d-531da2fd3686" />
<img width="700" height="1521" alt="image" src="https://github.com/user-attachments/assets/d22b46ef-cce7-48d7-90c4-13b749b7d974" />

**4. Addition of the Global Preferences Interface (Fixes #7442)**  
A new Global Preferences page enables Institution Admins to configure database-wide settings, including auditing, date formats, and thumbnail sizes. This UI replaces manual preference edits and is backed by new API endpoints for retrieving and updating global preference values.

Screenshots:
<img width="300" height="836" alt="image" src="https://github.com/user-attachments/assets/eb1498ee-6c13-464b-97ae-4660e9c7fc8d" />
<img width="700" height="2089" alt="image" src="https://github.com/user-attachments/assets/a365b2ed-aa38-4086-964b-c99af32f91e6" />



### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good and
      self-explanatory (or properly documented)
- [ ] Add relevant issue to release milestone

### Testing instructions
###  Retire obsolete remotePrefs and migrate Tree Viewer to User Preferences

- [ ] Verify Tree Viewer functionality works correctly.
- [ ] Load a tree view (e.g., Taxon) with the network panel open; Verify that expanding a node below that rank does trigger a /stats/ request and shows node statistics, while expanding a node above that rank does not
- [ ] Confirm the app loads without errors and obsolete are not used.

---

### Migrate Rank.Threshold to User Preferences

- [ ] Go to User Preferences
- [ ] Verify that under each tree there is an option to configure the "Show object count"

    <img width="595" alt="image" src="https://github.com/user-attachments/assets/d27e4094-d8da-42ae-8735-784737481427" />

- [ ] Verify that the ranks returned are correct for the current collection's tree(s)
- [ ] Select a rank from each tree
- [ ] Verify that the counts for all nodes under that level are shown in the tree
- [ ] Test this across all current trees and verify the behavior

---

### Feat: Add “Collection Preferences” page with role-gated access

- [ ] Verify the new Collection Preferences menu is visible only to authorized roles, settings can be saved and persist per collection.
- [ ] Modify values in each section, ensure they save and persist after reload, and verify that the Reset button restores defaults correctly.
- [ ] Test each option and verify application behavior responds to the setting

#### General
- [ ] Scope "Entire Table" Picklists
- [ ] Make Attachments Public By Default

#### Tree Management
If enabled, this allows users to add children to synonymized parents and to synonymize a node with children.
- [ ] Taxon
- [ ] Geography
- [ ] Storage
- [ ] Chronostratigraphy
- [ ] Lithostratigraphy
- [ ] Tectonic Unit

#### Statistics
- [ ] Show Preparation Totals
- [ ] Auto-Refresh Rate (Hours)
- [ ] GBIF Publishing Organization Key
- [ ] GBIF Data Set Key

#### Catalog Number Inheritance
- [ ] Enable Catalog number Inheritance From Primary Collection Object
- [ ] Enable Catalog number Inheritance From Parent Collection Object

---

### PR Feat: Refactor and add UI to control Global preferences

- [ ] Verify that the new Global Preferences page is accessible only to Institution Admins, settings (auditing, date formats, thumbnails) can be updated, saved, and persist correctly across reloads.
- [ ] Change and verify that all configuration options are applied

Gitesh307 committed:
Fix:Expose general collection prefs for picklist scoping and attachment defaults

Gitesh307 committed:
implement Collection Preferences UI with dedicated definitions

Gitesh307 committed:
adding missing files for the Collection preferences

Gitesh307 committed:
resolve typecheck errors in Collection Preferences definitions

Gitesh307 committed:
add Collection Preferences UI and editor with doc links

Gitesh307 committed:
implify CollectionPreferences page by removing internal save

Gitesh307 committed:
Corrected specifyNetwork keys to publishingOrg and datasetKey

Gitesh307 committed:
fix TypeScript errors in collection preference definitions

Gitesh307 committed:
align CollectionDefinitions with UserDefinitions typing

Gitesh307 committed:
Fix localization scanner issues in preferences

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by 20b207e0bd6648d0fd9991f97a1cca5eb0144998 on branch refs/heads/issue-7440

Gitesh307 committed:
adding path for role gated access

Gitesh307 committed:
Merge branch 'issue-7440' of https://github.com/specify/specify7 into issue-7440

Gitesh307 committed:
Refactor collection preferences renderer and documentation link layout

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by 78c6e544302cc9a05cc53aece3e7c53295e17944 on branch refs/heads/issue-7440

Gitesh307 committed:
Localization cleanup for collection preferences

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by c4ddea8efb09522d45738b84f5587708f324b757 on branch refs/heads/issue-7440

Gitesh307 committed:
removed collection pref refrence from User Tools

Gitesh307 committed:
remove extra language lines and keep only 'en-us'

Gitesh307 committed:
extract shared logic into createPreferencesEditor and wire User/Collection editors

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by ad70ddc29986383d6ead31e885fb1ccfb60458d2 on branch refs/heads/issue-7440

Gitesh307 committed:
implemented IR<T> for preferences typing

Gitesh307 committed:
Merge branch 'issue-7440' of https://github.com/specify/specify7 into issue-7440

Gitesh307 committed:
resolve BasePreferences context type incompatibility in index.tsx

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by e58d9f0d7089509cb0f0f5e0edb622beb2800281 on branch refs/heads/issue-7440

Gitesh307 committed:
simplify resolveCollectionDocumentHref logic using unified NAME_DOCS_MAP

Gitesh307 committed:
Merge branch 'issue-7440' of https://github.com/specify/specify7 into issue-7440

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by ff875da5c2871b33b2748d21ed233877702adb30 on branch refs/heads/issue-7440

Gitesh307 committed:
unify visibilityContext to remove duplication

Gitesh307 committed:
Merge branch 'issue-7440' of https://github.com/specify/specify7 into issue-7440

Gitesh307 committed:
optimize preferences wrappers by introducing FetchGate component.

Gitesh307 committed:
simplify className logic for documentation paragraph

Gitesh307 committed:
Localization cleanup for collection preferences

Gitesh307 committed:
removed `as unknown` casts

Gitesh307 committed:
consolidate doc links and merge catalog-number card

Gitesh307 committed:
removed unused imports

Gitesh307 committed:
function call

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by c69f079b8c6f12797890cc2450849d5e9d3d0711 on branch refs/heads/issue-7440

Gitesh307 committed:
render editor inline for User tools menu

Gitesh307 committed:
Merge branch 'issue-7440' of https://github.com/specify/specify7 into issue-7440

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by 04c17881a15b18e607c5554491a25aed4f56ca2c on branch refs/heads/issue-7440

Gitesh307 committed:
Feat: Refactor and add UI to control Global preferences

Gitesh307 committed:
Editor Configuration for Global Pref

Gitesh307 committed:
Global pref app resources integration

Gitesh307 committed:
localization strings

Gitesh307 committed:
Global Preferences View, Routing and User tools menu logic

Gitesh307 committed:
Add remotePreferences test coverage and stabilize ajax mocks

Gitesh307 committed:
ix test-mode ajax mock import without bundling node modules

Gitesh307 committed:
removed extra localization text

Gitesh307 committed:
removed redundant localization strings

Gitesh307 committed:
added the import statement

Gitesh307 committed:
Enable Global Preferences visual editor for legacy resources

Gitesh307 committed:
Fix Global Preferences tests and harden ajax mock

Gitesh307 committed:
Stabilize AppResources tests for Global Preferences rename

Gitesh307 committed:
declare explicit return type for flattenResources

Gitesh307 committed:
update AppResourcesTab and useResourcesTree test cases for Global Preferences

Gitesh307 committed:
remove unused container variable in AppResourcesTab.test.tsx

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by a0f9489c63bdd90ce80bb0dde5f5ad30e48bb23b on branch refs/heads/issue-7442-3

Gitesh307 committed:
restrict Global Preferences visual editor to 'Global Prefs' directory only

Gitesh307 committed:
Merge branch 'issue-7442-3' of https://github.com/specify/specify7 into issue-7442-3

Gitesh307 committed:
Fix Global Preferences date dropdowns in visual editor

Gitesh307 committed:
added localized helper to wrap the raw string in the LocalizedString type

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by e223eff34df542eb8bfe77e244c433f542b382ac on branch refs/heads/issue-7442-3

Gitesh307 committed:
Modularize preferences localization into smaller files

Gitesh307 committed:
createDictionary export for modularize prefrences

Gitesh307 committed:
Refactor preference localization dictionaries to expose raw maps

grantfitzsimmons committed:
fix: links to docs

Gitesh307 committed:
Modularize preferences localization dictionary

Gitesh307 committed:
Merge branch 'issue-7440' of https://github.com/specify/specify7 into issue-7440

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by b1b03396ac63b5e308ecd46143c8c30267665cc5 on branch refs/heads/issue-7440

grantfitzsimmons committed:
refactor(preferences): improve tree management

grantfitzsimmons committed:
fix: improve public attachment description

grantfitzsimmons committed:
Update attachments.ts

grantfitzsimmons committed:
feat: add descriptions for specify network fields

grantfitzsimmons committed:
feat: improve descriptions for stats fields

grantfitzsimmons committed:
feat: improve title for synonym behavior section

grantfitzsimmons committed:
feat: improve cat number inheritance

grantfitzsimmons committed:
feat: improve descriptions for catalog number inheritance

grantfitzsimmons committed:
feat: add better icon, distinguish preferences

grantfitzsimmons committed:
fix: reorder trees in order of relevance

grantfitzsimmons committed:
feat(collection preferences): add sidebar

grantfitzsimmons committed:
feat(preferences): match user preferences visual

grantfitzsimmons committed:
Merge branch 'main' into issue-7440

grantfitzsimmons committed:
fix: failing test

grantfitzsimmons committed:
fix: remove unused localization strings

grantfitzsimmons committed:
refactor: simplify code

grantfitzsimmons committed:
Lint code with ESLint and Prettier

Triggered by a0e546302ba2c795fa9e3dbe5fcbb21ff0049222 on branch refs/heads/issue-7440

grantfitzsimmons committed:
Merge branch 'issue-7440' into issue-7442-3

grantfitzsimmons committed:
fix: add missing localization strings

grantfitzsimmons committed:
feat(preferences): match collection preferences visual

grantfitzsimmons committed:
fix: add static test globalpreferences

grantfitzsimmons committed:
Lint code with ESLint and Prettier

Triggered by 914d3918f21d776eae491b4c3ad24969f73aa5e4 on branch refs/heads/issue-7442-3

grantfitzsimmons committed:
feat: organize global settings into categories

grantfitzsimmons committed:
Merge branch 'issue-7442-3' of https://github.com/specify/specify7 into issue-7442-3

grantfitzsimmons reviewed (changes requested):
Nice work! So much progress!

There are two global preference files, named `preferences`, denoted by different headers:

```
#Remote User Prefs
#Wed Sep 27 17:21:50 CEST 2023
```
and 
```
#Global Prefs
#Thu Jul 20 14:56:10 CEST 2023
```

This PR assumes that all preferences are in the global prefs resource, but some settings (like `ui.formatting.scrdateformat`) are only looked at in remote prefs. Users should not have their current date format lost when we release this, so we need to transition existing configurations if at all possible. 

Right now, despite this being set, queries and other places do not look to this resource for the date format

grantfitzsimmons committed:
fix: failing test

Gitesh307 committed:
fixing failing tests

Gitesh307 committed:
Migrate screen date format from remote prefs into GlobalPreferences

Gitesh307 committed:
resolve readonly assignment errors in global preferences utils and loader

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by 4d49db9c76d2a036d2b0b16e2179b741c210856a on branch refs/heads/issue-7442-3

Gitesh307 committed:
ensure Global Preferences read/write the global app resource

Gitesh307 committed:
satisfy TS readonly types in global preferences flow

Gitesh307 committed:
align tests with normalized formats and mock headers

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by d710796345ceb9c6ffeedccdd88e3b0bb318e742 on branch refs/heads/issue-7442-3

Gitesh307 committed:
Merge branch 'issue-7440' into issue-7442-3

Gitesh307 committed:
Fix remote pref definitions regression

Gitesh307 committed:
Fix collection pref usage and clean duplicate localization keys

Gitesh307 committed:
resolved merge conflicts

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by 1a9946b11629c3da6d96e8af85e1491a9ab715ee on branch refs/heads/issue-7442-3

Iwantexpresso reviewed (approved):
- [x] Verify that the new Global Preferences page is accessible only to Institution Admins, settings (auditing, date formats, thumbnails) can be updated, saved, and persist correctly across reloads.

- [x]  Change and verify that all configuration options are applied

-----
just tested on Calvert marine museum! these look good!

Gitesh307 committed:
Merge branch 'issue-7440' into issue-7442-3

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by 860f25b7bf5b60a9b820b10b7f43c06c2f8212fe on branch refs/heads/issue-7442-3

Gitesh307 committed:
Merge remote-tracking branch 'origin/issue-7440' into issue-7442-3

Gitesh307 committed:
Restrict global pref accessibilty to authorised roles

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by 88a5a2197293dc3be04947f5c9c7aa81e10c54e8 on branch refs/heads/issue-7442-3

emenslin reviewed (changes requested):
- [ ] Verify that the new Global Preferences page is accessible only to Institution Admins, settings (auditing, date formats, thumbnails) can be updated, saved, and persist correctly across reloads.
- [ ] Change and verify that all configuration options are applied

When any change is made to global preferences after clearing cache the results in global preferences don't seem to change (i.e. if you changed the date format and saved, it would still show the first date format), but sometimes it does show the changed results in queries and sometimes it doesn't, I'm not sure how recreate it one vs the other. Also either way it seems like the month/year date format preference is not respected. 


This shows that the query results were updated to the changes, but global preferences now does not match the query results. 

https://github.com/user-attachments/assets/3e135061-2fbb-4f7f-9f1c-1ebdd16a90b5

Gitesh307 committed:
Restore GlobalPreferences fetch to cache-buster list

Gitesh307 committed:
Handle undefined admin flag when filtering global preferences

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by e6687cdccf62e9d155ce267271dde12c0bb800eb on branch refs/heads/issue-7442-3

Gitesh307 committed:
Merge branch 'issue-7440' into issue-7442-3

emenslin reviewed (changes requested):
- [ ] Verify that the new Global Preferences page is accessible only to Institution Admins, settings (auditing, date formats, thumbnails) can be updated, saved, and persist correctly across reloads.
- [ ] Change and verify that all configuration options are applied

Partial date format is still not respected, and month/year dates just match the full date format. If you turn off accessible month format in user preferences you can see it is displayed properly in the forms but in queries it is not. 

https://github.com/user-attachments/assets/4f4a0a37-4668-4971-9bb5-8f61cc20673d


Disabling audit log and log field-level changes doesn't work

https://github.com/user-attachments/assets/e64900ed-03b2-46fc-823b-b5b69f101fc6


Default attachment size is also broken in this branch. Some attachments seem to overlap others. In the screenshots both the issue branch and main are using the same 256px thumbnail size. 

Issue branch:
<img width="1919" height="952" alt="image" src="https://github.com/user-attachments/assets/744c113f-d191-476b-9b33-23acf61b5577" />

Main: 
<img width="1919" height="949" alt="image" src="https://github.com/user-attachments/assets/8c4f9c70-0e2e-4538-9666-f8c37ed767bc" />

Gitesh307 committed:
Handle undefined admin flag when filtering global preferences

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by edf09cab33d2b182a5c0552b2908695420700d93 on branch refs/heads/issue-7442-3

Gitesh307 committed:
Merge branch 'issue-7440' into issue-7442-3

Gitesh307 committed:
Fix global prefs visibility when isadmin is undefined

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by 9694d44b3e144df0ab0d804f671ec742a9e22d02 on branch refs/heads/issue-7442-3

Gitesh307 committed:
gate stats fetch by rank preference

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by ab0ea15a3ef11dcc1aa8b9445062210c19cfe526 on branch refs/heads/issue-7442-3

Gitesh307 committed:
removed stats fetch by rank pref logic

Gitesh307 committed:
Fix audit log preference handling and add tests for toggle

Iwantexpresso reviewed (changes requested):
### Testing instructions
###  Retire obsolete remotePrefs and migrate Tree Viewer to User Preferences

- [x] Verify Tree Viewer functionality works correctly.
- [x] Load a tree view (e.g., Taxon) with the network panel open; expand a node whose rank is below the current threshold and confirm no …/stats/ request is issued, then expand a higher-rank node and confirm one is.
- [x] Confirm the app loads without errors and obsolete prefs like form.definition.columnSource are not used.

---

### Migrate Rank.Threshold to User Preferences

- [ ] Go to User Preferences
- [ ] Verify that under each tree there is an option to configure the "Show object count"

    <img width="595" alt="image" src="https://github.com/user-attachments/assets/d27e4094-d8da-42ae-8735-784737481427" />

- [x] Verify that the ranks returned are correct for the current collection's tree(s)
- [x] Select a rank from each tree
- [ ] Verify that the counts for all nodes under that level are shown in the tree
- [ ] Test this across all current trees and verify the behavior

 Counts are not being displayed at all, tested with geography and taxon tree on CalvertMarinemuseum with both light and dark mode. Tried clearing cache after saving changes as well as logging in and out and the result was still the same.
<img width="1448" height="734" alt="image" src="https://github.com/user-attachments/assets/7e3956ea-0f17-4e82-9a5d-a632b97ca006" />
<img width="1578" height="1191" alt="image" src="https://github.com/user-attachments/assets/761f3cf8-2951-4f59-aa12-a5ef4b7d63eb" />
------------------

Gitesh307 committed:
statsThreshold logic correction and auditlog tests

Gitesh307 committed:
Respect global attachment thumbnail preference across attachment views

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by 4083942017047b6741d62cdd4625e3f51fe891df on branch refs/heads/issue-7442-3

Gitesh307 committed:
Drop obsolete entries from remote prefs which is now handled by Global prefs

Gitesh307 committed:
Sync Global Preferences between user tools and App Resources

Gitesh307 committed:
Lint code with ESLint and Prettier

Triggered by c8594cba094e71c959685edab14c07c8ad9d8fd2 on branch refs/heads/issue-7442-3

Gitesh307 committed:
resolved multiobject returned error for sync global prefs

Gitesh307 committed:
Add shared notifier for global preference json and UI sync

Gitesh307 committed:
Revise .env file with database user details

Updated database user configurations and added comments for clarity.
